### PR TITLE
Fix completions for contextually constrained types in JSDoc

### LINKED ISSUE #3010: nil pointer dereference for contextual constraint completions in JSDoc
```ts
// @allowJs: true
// @filename: a.ts
export interface Blah<T extends { a: "hello" | "world" }> {
}

// @filename: b.js
/** @import * as a from "./a" */

/** @type {a.Blah<{ a: /*1*/ }>} */
let x;

// @filename: c.js
/** @import * as a from "./a" */

/** @type {a.Blah<{ a: /*2*/ }>} */
```

Request completions at either marker:

```
internal/ls.getConstraintOfTypeArgumentProperty()
	internal/ls/completions.go:3658
internal/ls.getConstraintOfTypeArgumentProperty()
	internal/ls/completions.go:3651
internal/ls.(*LanguageService).getCompletionData()
	internal/ls/completions.go:1604
internal/ls.(*LanguageService).getCompletionsAtPosition()
	internal/ls/completions.go:347
internal/ls.(*LanguageService).ProvideCompletion()
	internal/ls/completions.go:47
internal/lsp.(*Server).handleCompletion()
	internal/lsp/server.go:1197
internal/lsp.init.func1.registerLanguageServiceWithAutoImportsRequestHandler[...].28.1()
	internal/lsp/server.go:775
internal/lsp.(*Server).handleRequestOrNotification.func1()
	internal/lsp/server.go:613
internal/lsp.(*Server).dispatchLoop.func3()
	internal/lsp/server.go:488
internal/lsp.(*Server).dispatchLoop
	internal/lsp/server.go:487
```

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
Basically we had a missing symbol because we were navigating the "original" syntax tree, and we were not grabbing the reparsed node. Unfortunately, that's not sufficient because not every JSDoc type node will have a corresponding reparsed node - in some cases, JSDoc nodes will never be re-attached to the resulting tree.

The code is resilient to these edge-cases and checks for an actual symbol, but tries to use the name of a syntax node as a fallback.

Fixes #3010.

### CODE REVIEW

File: internal/fourslash/tests/completionsForContextualConstraintTypeInJsDoc_test.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+	// because of how JSDoc re-parsing would construct nodes in the tree.
+	//
+	// Getting the symbol of the reparsed node is a sufficient fix for marker 1.
+	// However, that would not fix the case at marker 2 because
+	// there is no variable to attach the `@type` annotation, so the node basically
+	// doesn't exist for subsequent passes like the binder.
+
+	f.VerifyCompletions(t, f.Markers(), &fourslash.CompletionsExpectedList{
+		IsIncomplete: false,
+		ItemDefaults: &fourslash.CompletionsExpectedItemDefaults{CommitCharacters: &[]string{".", ",", ";"}},
```
[the remark this answers is not in the document]
Comment (DanielRosenwasser) [reply internal/fourslash/tests/completionsForContextualConstraintTypeInJsDoc_test.go:50]: Hard failure? What?

I do wonder if we need to specify any of these `ItemDefaults` though.
